usual/tls/tls.c

```diff
@@ -742,4 +742,85 @@ tls_close(struct tls *ctx)
 	return (rv);
 }
 
+static bool
+tls_mems_equal(char *mem1, char *mem2, size_t len1, size_t len2)
+{
+	if (len1 != len2)
+		return false;
+	if (mem1 && mem2 && memcmp(mem1, mem2, len1) != 0)
+		return false;
+	return true;
+}
+
+static bool
+keypair_equal(struct tls_keypair *tkp1, struct tls_keypair *tkp2)
+{
+	if (!strcmpeq(tkp1->cert_file, tkp2->cert_file))
+		return false;
+	if (!tls_mems_equal(tkp1->cert_mem, tkp2->cert_mem, tkp1->cert_len, tkp2->cert_len))
+		return false;
+	if (!strcmpeq(tkp1->key_file, tkp2->key_file))
+		return false;
+	if (!tls_mems_equal(tkp1->key_mem, tkp2->key_mem, tkp1->key_len, tkp2->key_len))
+		return false;
+	return true;
+}
+
+bool
+keypair_list_equal(struct tls_keypair *tkp1, struct tls_keypair *tkp2)
+{
+	bool kp_unchanged;
+	struct tls_keypair *tls_kp_lt, *tls_kp_rt;
+
+	kp_unchanged = keypair_equal(tkp1, tkp2);
+
+	if (kp_unchanged == false) {
+		return false;
+	}
+
+	for (tls_kp_lt = tkp1, tls_kp_rt = tkp2 ; tls_kp_lt != NULL && tls_kp_rt != NULL; tls_kp_lt = tls_kp_lt->next, tls_kp_rt = tls_kp_rt->next) {
+		kp_unchanged = keypair_equal(tls_kp_lt, tls_kp_rt);
+	}
+
+	return kp_unchanged;
```

Comment:
I think this whole function can be simplified to this. And we should also check if the lists are of the same length.
```suggestion
	for (; tkp1 == || tkp2 == NULL; tkp1 = tkp1->next, tkp2 = tkp2->next) {
		if (!keypair_equal(tpk1, tkp2))
			return false;
	}
	// Check if the length of the lists are different
	return tkp1 != NULL || tkp2 != NULL;
```

Reply:
Thanks! This really is a _lot_ simpler. Additionally, I found it helpful to write a new test to drive out the boolean logic. [5467377](https://github.com/libusual/libusual/pull/57/commits/5467377ef1c17263ebc385e681ccd2addc5ca308)